### PR TITLE
Interact with the namespace selector

### LINKED ISSUE #5994: Namespaced menu item can't interact with the namespace selector
/kind feature

**Why you need this feature:**
A new feature to add namespaced menu items was merged by #5871, it can't interact with the namespace selector.

**[Example]**
Assuming MLFlow is added as `/mlflow/{ns}` to the menubar, and namespace=workspace1 is activated, the full URL would be ` https://gateway/_/mlflow/workspace1/?ns=workspace1`.
Accessing that page and changing the namespace to workspace2 by the namespace selector, the URL become inconsistent as `https://gateway/_/mlflow/workspace1/?ns=workspace2` and no menu item is selected.

**Describe the solution you'd like:**
When the user control the namespace selector, namespaced menu items should follow it.

In a below example, ` https://gateway/_/mlflow/workspace1/?ns=workspace1` should move to `https://gateway/_/mlflow/workspace2/?ns=workspace2`, if the namespaced is changed from workspace1 to workspace2.

**Anything else you would like to add:**
* This feature is explained in the below page.<br>https://www.kubeflow.org/docs/components/central-dash/customizing-menu/
* I'm going to send a PR to solve this problem.

/cc @kimwnasptd

### PR DESCRIPTION
**Pull request by toshi-k**
Fix
#5994 

This PR enables namespaced menu items to interact with the namespace selector.
It means the URL of namespaced menu items will follow the selection of namespace selector.

The feature of namespaced menu items is merged by #5871, and described in [here](https://www.kubeflow.org/docs/components/central-dash/customizing-menu/) .

When the user chooses namespace=workspace1, the URL well be `https://gateway/_/mlflow/workspace1/?ns=workspace1`. If he chooses namespace=workspace2 later, the URL will change into `https://gateway/_/mlflow/workspace2/?ns=workspace2` dynamically.

/cc @kimwnasptd @DavidSpek

### CODE REVIEW

**Comment by karlschriek** [comment]
Hey @toshi-k, @kimwnasptd @DavidSpek. Looking at this as well as at https://github.com/kubeflow/kubeflow/pull/5871, I am wondering whether such filtering on namespace level couldn't be achieved more simply by using `queryParams` on the VirtualService rather than by building custom logic into the central dashboard application. 

Let's say for example we want to roll out the MLFlow Server in every profile Namespace, couldn't achieve this as follows without needing these two PRs?

```
kind: VirtualService
apiVersion: networking.istio.io/v1alpha3
metadata:  
  name: mlflow
  namespace: karl-schriek
spec:
  gateways:
    - kubeflow/kubeflow-gateway
  hosts:
    - '*'
  http:
    - match:
        - uri:
            prefix: /mlflow
          queryParams:
            ns:
              exact:  karl-schriek
      rewrite:
        uri: /
      route:
        - destination:
            host: mlflow.karl-schriek.svc.cluster.local
            port:
              number: 80
---
apiVersion: v1
kind: Service
metadata:
  name: mlflow
  namespace: karl-schriek
spec:
  selector:
    app: mlflow
  type: ClusterIP
  ports:
  - port: 80
    targetPort: 5000
```

I quickly hacked something together on this today. Haven't managed to get it working (even without the "queryParams" part I am not able to get it to work), but I think in theory it should be possible. Are you aware of anything that would prevent an implementation like this from working?

**Comment by kimwnasptd** [comment]
Thank you for the contribution @toshi-k, the Dashboard changes for MLFlow are really looking good!

I'll take a more detailed look at this within the next couple of days.

@karlschriek IIUC we had also discussed about this alternative for the first PR in https://github.com/kubeflow/kubeflow/issues/5863#issuecomment-847156432, but in the end as @toshi-k mentioned the query param gets only set the first time the user navigates to the app. 

For example, let say that the namespaced link was a grafana instance in a specific namespace. Once the user would navigate to it via the Central Dashboard then the request would contain the `?ns={namespace}` query parameter. All good. But then the frontend code of Grafana would not be adding the `ns={namsepace}` query params in the requests it makes, since it doesn't know anything about it.

@karlschriek did you have something else in mind?

**Comment by karlschriek** [comment]
Ok, interesting. So if you click on "Grafana" in the kubeflow dashboard left-hand panel it works, but then inside the iframed Grafana dashboard if you navigate its subpages, the `?ns={namespace}` query parameter gets lost? Or is it the case of not being able to navigate from "Grafana" to "Notebooks" for example?

But @toshi-k's PR here addresses exactly that problem, correct? 

I didn't have anything else specific in mind, I just always prefer a solution that doesn't rely on application-level changes if it is on offer.

**Comment by toshi-k** [comment]
@kimwnasptd Thank you ! I'm looking for your feedback.

@karlschriek I thought the same as you initially, but there are many issues to control traffic by queryParam. Although Kubeflow 1.3 is getting better to keep ?ns=namespace superficially, it doesn't work well. I think one of the problems is that ns parameter is deleted for the inside content of iframe.  
```
   _setIframeSrc() {
        const iframeUrl = new URL(this.subRouteData.path,
            window.location.origin);
        iframeUrl.hash = window.location.hash;
        iframeUrl.search = window.location.search;
        iframeUrl.searchParams.delete('ns');
        this.iframeSrc = iframeUrl.toString();
    }
```
https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/public/components/main-page.js#L311

I commented out `iframeUrl.searchParams.delete('ns')` at once, it caused side effects and failed to pass the tests.

Another concern is that some applications need additional API accesses between the app server and the client (UI). queryParam would not be added to all API accesses and it would cause errors.

As I said in #5863 , I concluded that path parameter is the most stable way to access third party app through kubeflow dashboard. If you find a perfect solution realized by queryParam, please share it in the community.

**Comment by toshi-k** [comment]
@kimwnasptd My PR works perfectly for MLFlow as the attached image !

mlflow

**Comment by toshi-k** [comment]
Sorry, I found a kind of another issue. I'm going to investigate it.
/hold

**Comment by toshi-k** [comment]
OK, I found a solution ! This PR is ready to be merged.
/unhold

**Comment by toshi-k** [comment]
@kimwnasptd You can use below manifests to deploy MLFlow temporarily for ns=workspace1 !
```
apiVersion: v1
kind: Service
metadata:
  name: mlflow
  namespace: mlflow
spec:
  ports:
  - name: http
    port: 5000
    targetPort: 5000
  selector:
    app: mlflow
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mlflow-deployment
  namespace: mlflow
spec:
  replicas: 1
  selector:
    matchLabels:
      app: mlflow
  template:
    metadata:
      labels:
        app: mlflow
    spec:
      containers:
      - command:
        - mlflow
        - server
        - --host
        - 0.0.0.0
        - --static-prefix
        - /mlflow/workspace1
        image: larribas/mlflow:latest
        imagePullPolicy: Always
        name: mlflow
        ports:
        - containerPort: 5000
          name: http
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: mlflow-vs
  namespace: mlflow
spec:
  hosts:
  - "*"
  gateways:
  - kubeflow/kubeflow-gateway
  http:
  - match:
    - uri:
        prefix: /mlflow/workspace1
    route:
    - destination:
        port:
          number: 5000
        host: mlflow
```
MLFlow URL from the central dashboard would be https://gateway/_/mlflow/workspace1/?ns=workspace1.

**Comment by kimwnasptd** [comment]
@toshi-k apologies for the delayed review on this and thank you very much for the YAMLs!

Taking a look at the PR

File: components/centraldashboard/public/components/main-page.js [hunk 1]
```diff
@@ -106,6 +106,8 @@ export class MainPage extends utilitiesMixin(PolymerElement) {
                 // eslint-disable-next-line max-len
                 computed: 'computeShouldFetchEnv(registrationFlow, workgroupStatusHasLoaded)',
             },
+            matchingIndex: Number,
```
Comment (kimwnasptd) [inline components/centraldashboard/public/components/main-page.js:109]: Why did you need to add this to the properties? IIUC it is only used from the `_setActiveMenuLink` method

**Review by kimwnasptd — COMMENTED** [review]
@toshi-k I also tested the app and it works as expected, nice work!

File: components/centraldashboard/public/components/main-page.js [hunk 2]
```diff
@@ -360,19 +374,23 @@ export class MainPage extends utilitiesMixin(PolymerElement) {
         });
         if (hashPath) {
             matchPath = path + '#' + hashPath;
-            matchingLink = allLinks
-                .find((l) => this.compareLinks(l, matchPath));
+            this.matchingIndex = allLinks
+                .findIndex((l) => this.compareLinks(l, matchPath));
         } else {
```
Comment (toshi-k) [inline components/centraldashboard/public/components/main-page.js:379]: @kimwnasptd Good point ! When the namespace is changed by the namespace selector, both current URL and menu links are updated. If current URL is updated earlier, the new URL(e.g /mlflow/workspace1) can't match with the prefix of old menu links (e.g /mlflow/workspace2).

In that case, all `path.startsWith(link)` may be False, and matchingIndex is not updated. Making it class variable `this.matchingIndex`, UI can remember which menu item is selected, and re-link the URL and menu-bar.

File: components/centraldashboard/public/components/main-page.js (hunk 1 above)
Comment (toshi-k) [reply components/centraldashboard/public/components/main-page.js:109]: I answerer that question in a below window.

File: components/centraldashboard/public/components/main-page.js (hunk 2 above)
Comment (kimwnasptd) [reply components/centraldashboard/public/components/main-page.js:379]: Makes sense, thanks for the explanation!

**Comment by kimwnasptd** [comment]
@toshi-k thank you for your work! I've also tested this locally and in my cluster and works as expected.

/lgtm
/approve

**Comment by toshi-k** [comment]
@kimwnasptd Thank you for your review !

I rebased my branch and updated this PR by below commands.
```
git pull origin interact-with-ns-selector --rebase
git push origin interact-with-ns-selector
```
But it seems that another problem occurred. Did I made a mistake ?
I appreciate if you give me some instructions.

**Comment by kimwnasptd** [comment]
It seems that the rebase included other commits as well. Could you do the following:
```bash
# Assuming that "public" is the kubeflow/kubeflow git repo
# Bring you local master up to date with upstream master
git checkout master
git pull public master

# rebase only your commits on top of the latest master
git checkout interact-with-ns-selector
git rebase -i master

# In the editor change the 'pick' word to 'drop' for any commit that you don't want.
# You should only select your three commits: 
# Interact with the namespace selector
# Use namespace given as an argument
# Keep the index of selected item

# finally force push your branch
git push origin HEAD -f
```

**Comment by toshi-k** [comment]
@kimwnasptd Thank you for your advice !
I modified my PR by force push.
I think this PR has only my 3 commits and my remote repo is at the latest master branch.
https://github.com/toshi-k/kubeflow/commits/interact-with-ns-selector
